
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.1.0">
    
    
      
        <title>External Exome - Workflow Launcher</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.bc7e593a.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab28b872.min.css">
        
          
          
          <meta name="theme-color" content="#4cae4f">
        
      
    
    
    
      
        
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="green" data-md-color-accent="teal">
      
  
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#external-exome-reprocessing-workload" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href=".." title="Workflow Launcher" class="md-header-nav__button md-logo" aria-label="Workflow Launcher">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      
        <div class="md-header-nav__ellipsis">
          <span class="md-header-nav__topic md-ellipsis">
            Workflow Launcher
          </span>
          <span class="md-header-nav__topic md-ellipsis">
            
              External Exome
            
          </span>
        </div>
      
    </div>
    
      <label class="md-header-nav__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active">
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" data-md-component="search-reset" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/broadinstitute/wfl/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Workflow Launcher" class="md-nav__button md-logo" aria-label="Workflow Launcher">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Workflow Launcher
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/broadinstitute/wfl/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." class="md-nav__link">
      Get Started
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2">
    
    <label class="md-nav__link" for="nav-2">
      Development Guide
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="Development Guide" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        <span class="md-nav__icon md-icon"></span>
        Development Guide
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../dev-process/" class="md-nav__link">
      Development Process and Tips
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../dev-sandbox/" class="md-nav__link">
      Local Development SandBox
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../dev-logging/" class="md-nav__link">
      Logging
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../dev-deployment/" class="md-nav__link">
      Deployment
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../dev-frontend/" class="md-nav__link">
      Frontend
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" checked>
    
    <label class="md-nav__link" for="nav-3">
      Modules
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="Modules" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        <span class="md-nav__icon md-icon"></span>
        Modules
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../modules-general/" class="md-nav__link">
      General
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../modules-arrays/" class="md-nav__link">
      Arrays
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../dev-wgs/" class="md-nav__link">
      Whole Genome
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        External Exome
        <span class="md-nav__icon md-icon"></span>
      </label>
    
    <a href="./" class="md-nav__link md-nav__link--active">
      External Exome
    </a>
    
      
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#inputs" class="md-nav__link">
    Inputs
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    Usage
  </a>
  
    <nav class="md-nav" aria-label="Usage">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#create-workload-apiv1create" class="md-nav__link">
    Create Workload: /api/v1/create
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#start-workload-apiv1start" class="md-nav__link">
    Start Workload: /api/v1/start
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exec-workload-apiv1exec" class="md-nav__link">
    Exec Workload: /api/v1/exec
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#query-workload-apiv1workloaduuiduuid" class="md-nav__link">
    Query Workload: /api/v1/workload?uuid=&lt;uuid&gt;
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#a1-external-exome-workload-request-json-spec" class="md-nav__link">
    A1 External Exome Workload-Request JSON Spec
  </a>
  
</li>
      
    </ul>
  
</nav>
    
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4">
    
    <label class="md-nav__link" for="nav-4">
      Readings
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="Readings" data-md-level="1">
      <label class="md-nav__title" for="nav-4">
        <span class="md-nav__icon md-icon"></span>
        Readings
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../terra/" class="md-nav__link">
      WorkFlow Launcher's Role in Terra
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#inputs" class="md-nav__link">
    Inputs
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    Usage
  </a>
  
    <nav class="md-nav" aria-label="Usage">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#create-workload-apiv1create" class="md-nav__link">
    Create Workload: /api/v1/create
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#start-workload-apiv1start" class="md-nav__link">
    Start Workload: /api/v1/start
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exec-workload-apiv1exec" class="md-nav__link">
    Exec Workload: /api/v1/exec
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#query-workload-apiv1workloaduuiduuid" class="md-nav__link">
    Query Workload: /api/v1/workload?uuid=&lt;uuid&gt;
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#a1-external-exome-workload-request-json-spec" class="md-nav__link">
    A1 External Exome Workload-Request JSON Spec
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/broadinstitute/wfl/edit/master/docs/external-exome-reprocessing.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="external-exome-reprocessing-workload">External Exome Reprocessing workload<a class="headerlink" href="#external-exome-reprocessing-workload" title="Permanent link">&para;</a></h1>
<h2 id="inputs">Inputs<a class="headerlink" href="#inputs" title="Permanent link">&para;</a></h2>
<p>An <code>ExternalExomeReprocessing</code> workload requires the specification of exactly
one the following inputs for each workflow:</p>
<ul>
<li><code>input_bam</code>, OR</li>
<li><code>input_cram</code></li>
</ul>
<p>Where <code>input_bam</code> and <code>input_cram</code> are GS URLs of the file to reprocess. Note
that the <code>input_bam</code> and <code>input_cram</code> inputs should only be used with CRAM and
BAM files, respectively. All other WDL inputs are optional - see the output 
below for all options.</p>
<h2 id="usage">Usage<a class="headerlink" href="#usage" title="Permanent link">&para;</a></h2>
<h3 id="create-workload-apiv1create">Create Workload: <code>/api/v1/create</code><a class="headerlink" href="#create-workload-apiv1create" title="Permanent link">&para;</a></h3>
<p>Create a new workload. Ensure that <code>workflow-launcher</code> and <code>cromwell</code>'s service
accounts have at least read access to the input files.</p>
<p>Request:
<div class="highlight"><pre><span></span><code>curl -X POST &#39;https://dev-wfl.gotc-dev.broadinstitute.org/api/v1/create&#39; \
  -H &quot;Authorization: Bearer $(gcloud auth print-access-token)&quot; \
  -H &#39;Content-Type: application/json&#39; \
  -d &#39;{
        &quot;cromwell&quot;: &quot;https://cromwell-gotc-auth.gotc-dev.broadinstitute.org&quot;,
        &quot;output&quot;: &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/&quot;,
        &quot;pipeline&quot;: &quot;ExternalExomeReprocessing&quot;,
        &quot;project&quot;: &quot;PO-1234&quot;,
        &quot;items&quot;: [{
            &quot;inputs&quot;: {
                &quot;input_cram&quot;: &quot;gs://path/to/a.cram&quot;
            }
        }]
      }&#39;
</code></pre></div>
Response:
<div class="highlight"><pre><span></span><code>{
  &quot;creator&quot; : &quot;user@domain&quot;,
  &quot;pipeline&quot; : &quot;ExternalExomeReprocessing&quot;,
  &quot;cromwell&quot; : &quot;https://cromwell-gotc-auth.gotc-dev.broadinstitute.org&quot;,
  &quot;release&quot; : &quot;ExternalExomeReprocessing_vX.Y.Z&quot;,
  &quot;created&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
  &quot;output&quot; : &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/&quot;,
  &quot;workflows&quot; : [ {
    &quot;inputs&quot; : {
      &quot;cram_ref_fasta&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta&quot;,
      &quot;scatter_settings&quot; : {
        &quot;haplotype_scatter_count&quot; : 50,
        &quot;break_bands_at_multiples_of&quot; : 0
      },
      &quot;input_cram&quot; : &quot;gs://path/to/a.cram&quot;,
      &quot;destination_cloud_path&quot; : &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/to&quot;,
      &quot;sample_name&quot; : &quot;a&quot;,
      &quot;bait_set_name&quot; : &quot;whole_exome_illumina_coding_v1&quot;,
      &quot;references&quot; : {
        &quot;calling_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/exome_calling_regions.v1.interval_list&quot;,
        &quot;contamination_sites_bed&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.bed&quot;,
        &quot;dbsnp_vcf_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf.idx&quot;,
        &quot;haplotype_database_file&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.haplotype_database.txt&quot;,
        &quot;known_indels_sites_indices&quot; : [ &quot;gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz.tbi&quot;, &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz.tbi&quot; ],
        &quot;evaluation_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/exome_evaluation_regions.v1.interval_list&quot;,
        &quot;contamination_sites_mu&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.mu&quot;,
        &quot;dbsnp_vcf&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf&quot;,
        &quot;known_indels_sites_vcfs&quot; : [ &quot;gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz&quot;, &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz&quot; ],
        &quot;reference_fasta&quot; : {
          &quot;ref_pac&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.pac&quot;,
          &quot;ref_bwt&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.bwt&quot;,
          &quot;ref_dict&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dict&quot;,
          &quot;ref_ann&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.ann&quot;,
          &quot;ref_fasta_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.fai&quot;,
          &quot;ref_alt&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.alt&quot;,
          &quot;ref_fasta&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta&quot;,
          &quot;ref_sa&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.sa&quot;,
          &quot;ref_amb&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.amb&quot;
        },
        &quot;contamination_sites_ud&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.UD&quot;
      },
      &quot;cram_ref_fasta_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.fai&quot;,
      &quot;papi_settings&quot; : {
        &quot;agg_preemptible_tries&quot; : 3,
        &quot;preemptible_tries&quot; : 3
      },
      &quot;base_file_name&quot; : &quot;a.cram&quot;,
      &quot;bait_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/HybSelOligos/whole_exome_illumina_coding_v1/whole_exome_illumina_coding_v1.Homo_sapiens_assembly38.baits.interval_list&quot;,
      &quot;final_gvcf_base_name&quot; : &quot;a.cram&quot;,
      &quot;target_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/HybSelOligos/whole_exome_illumina_coding_v1/whole_exome_illumina_coding_v1.Homo_sapiens_assembly38.targets.interval_list&quot;,
      &quot;unmapped_bam_suffix&quot; : &quot;.unmapped.bam&quot;
    }
  } ],
  &quot;commit&quot; : &quot;commit-ish&quot;,
  &quot;project&quot; : &quot;PO-1234&quot;,  
  &quot;uuid&quot; : &quot;1337254e-f7d8-438d-a2b3-a74b199fee3c&quot;,  
  &quot;wdl&quot; : &quot;pipelines/broad/reprocessing/external/exome/ExternalExomeReprocessing.wdl&quot;,
  &quot;version&quot; : &quot;X.Y.Z&quot;
}
</code></pre></div></p>
<h3 id="start-workload-apiv1start">Start Workload: <code>/api/v1/start</code><a class="headerlink" href="#start-workload-apiv1start" title="Permanent link">&para;</a></h3>
<p>Starts a Cromwell workflow for each item in the workload. If an output already exists in the output bucket for a
particular input cram, WFL will not re-submit that workflow.</p>
<p>Request:
<div class="highlight"><pre><span></span><code>curl -X POST &#39;https://dev-wfl.gotc-dev.broadinstitute.org/api/v1/start&#39; \
 -H &quot;Authorization: Bearer $(gcloud auth print-access-token)&quot; \
 -H &#39;Content-Type: application/json&#39; \
 -d &#39;[{&quot;uuid&quot;: &quot;1337254e-f7d8-438d-a2b3-a74b199fee3c&quot;}]&#39;
</code></pre></div>
Response:
<div class="highlight"><pre><span></span><code>[{
   &quot;creator&quot; : &quot;user@domain&quot;,
   &quot;pipeline&quot; : &quot;ExternalExomeReprocessing&quot;,
   &quot;cromwell&quot; : &quot;https://cromwell-gotc-auth.gotc-dev.broadinstitute.org&quot;,
   &quot;release&quot; : &quot;ExternalExomeReprocessing_vX.Y.Z&quot;,
   &quot;created&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
   &quot;started&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
   &quot;output&quot; : &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/&quot;,
   &quot;workflows&quot; : [ {
     &quot;status&quot; : &quot;Submitted&quot;
     &quot;updated&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
     &quot;uuid&quot; : &quot;bb0d93e3-1a6a-4816-82d9-713fa58fb235&quot;,
     &quot;inputs&quot; : {
       &quot;cram_ref_fasta&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta&quot;,
       &quot;scatter_settings&quot; : {
         &quot;haplotype_scatter_count&quot; : 50,
         &quot;break_bands_at_multiples_of&quot; : 0
       },
       &quot;input_cram&quot; : &quot;gs://path/to/a.cram&quot;,
       &quot;destination_cloud_path&quot; : &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/to&quot;,
       &quot;sample_name&quot; : &quot;a&quot;,
       &quot;bait_set_name&quot; : &quot;whole_exome_illumina_coding_v1&quot;,
       &quot;references&quot; : {
         &quot;calling_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/exome_calling_regions.v1.interval_list&quot;,
         &quot;contamination_sites_bed&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.bed&quot;,
         &quot;dbsnp_vcf_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf.idx&quot;,
         &quot;haplotype_database_file&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.haplotype_database.txt&quot;,
         &quot;known_indels_sites_indices&quot; : [ &quot;gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz.tbi&quot;, &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz.tbi&quot; ],
         &quot;evaluation_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/exome_evaluation_regions.v1.interval_list&quot;,
         &quot;contamination_sites_mu&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.mu&quot;,
         &quot;dbsnp_vcf&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf&quot;,
         &quot;known_indels_sites_vcfs&quot; : [ &quot;gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz&quot;, &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz&quot; ],
         &quot;reference_fasta&quot; : {
           &quot;ref_pac&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.pac&quot;,
           &quot;ref_bwt&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.bwt&quot;,
           &quot;ref_dict&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dict&quot;,
           &quot;ref_ann&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.ann&quot;,
           &quot;ref_fasta_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.fai&quot;,
           &quot;ref_alt&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.alt&quot;,
           &quot;ref_fasta&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta&quot;,
           &quot;ref_sa&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.sa&quot;,
           &quot;ref_amb&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.amb&quot;
         },
         &quot;contamination_sites_ud&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.UD&quot;
       },
       &quot;cram_ref_fasta_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.fai&quot;,
       &quot;papi_settings&quot; : {
         &quot;agg_preemptible_tries&quot; : 3,
         &quot;preemptible_tries&quot; : 3
       },
       &quot;base_file_name&quot; : &quot;a.cram&quot;,
       &quot;bait_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/HybSelOligos/whole_exome_illumina_coding_v1/whole_exome_illumina_coding_v1.Homo_sapiens_assembly38.baits.interval_list&quot;,
       &quot;final_gvcf_base_name&quot; : &quot;a.cram&quot;,
       &quot;target_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/HybSelOligos/whole_exome_illumina_coding_v1/whole_exome_illumina_coding_v1.Homo_sapiens_assembly38.targets.interval_list&quot;,
       &quot;unmapped_bam_suffix&quot; : &quot;.unmapped.bam&quot;
     }
   } ],
   &quot;commit&quot; : &quot;commit-ish&quot;,
   &quot;project&quot; : &quot;PO-1234&quot;,  
   &quot;uuid&quot; : &quot;1337254e-f7d8-438d-a2b3-a74b199fee3c&quot;,  
   &quot;wdl&quot; : &quot;pipelines/broad/reprocessing/external/exome/ExternalExomeReprocessing.wdl&quot;,
   &quot;version&quot; : &quot;X.Y.Z&quot;
 }]
</code></pre></div></p>
<h3 id="exec-workload-apiv1exec">Exec Workload: <code>/api/v1/exec</code><a class="headerlink" href="#exec-workload-apiv1exec" title="Permanent link">&para;</a></h3>
<p>Creates and then starts a Cromwell workflow for each item in the workload.</p>
<p>Request:
<div class="highlight"><pre><span></span><code>curl -X POST &#39;https://dev-wfl.gotc-dev.broadinstitute.org/api/v1/exec&#39; \
  -H &quot;Authorization: Bearer $(gcloud auth print-access-token)&quot; \
  -H &#39;Content-Type: application/json&#39; \
  -d &#39;{
        &quot;cromwell&quot;: &quot;https://cromwell-gotc-auth.gotc-dev.broadinstitute.org&quot;,
        &quot;output&quot;: &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/&quot;,
        &quot;pipeline&quot;: &quot;ExternalExomeReprocessing&quot;,
        &quot;project&quot;: &quot;PO-1234&quot;,
        &quot;items&quot;: [{
            &quot;inputs&quot;: {
                &quot;input_cram&quot;: &quot;gs://path/to/a.cram&quot;
            }
        }]
      }&#39;
</code></pre></div>
Response:
<div class="highlight"><pre><span></span><code>{
   &quot;creator&quot; : &quot;user@domain&quot;,
   &quot;pipeline&quot; : &quot;ExternalExomeReprocessing&quot;,
   &quot;cromwell&quot; : &quot;https://cromwell-gotc-auth.gotc-dev.broadinstitute.org&quot;,
   &quot;release&quot; : &quot;ExternalExomeReprocessing_vX.Y.Z&quot;,
   &quot;created&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
   &quot;started&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
   &quot;output&quot; : &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/&quot;,
   &quot;workflows&quot; : [ {
     &quot;status&quot; : &quot;Submitted&quot;
     &quot;updated&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
     &quot;uuid&quot; : &quot;bb0d93e3-1a6a-4816-82d9-713fa58fb235&quot;,
     &quot;inputs&quot; : {
       &quot;cram_ref_fasta&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta&quot;,
       &quot;scatter_settings&quot; : {
         &quot;haplotype_scatter_count&quot; : 50,
         &quot;break_bands_at_multiples_of&quot; : 0
       },
       &quot;input_cram&quot; : &quot;gs://path/to/a.cram&quot;,
       &quot;destination_cloud_path&quot; : &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/to&quot;,
       &quot;sample_name&quot; : &quot;a&quot;,
       &quot;bait_set_name&quot; : &quot;whole_exome_illumina_coding_v1&quot;,
       &quot;references&quot; : {
         &quot;calling_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/exome_calling_regions.v1.interval_list&quot;,
         &quot;contamination_sites_bed&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.bed&quot;,
         &quot;dbsnp_vcf_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf.idx&quot;,
         &quot;haplotype_database_file&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.haplotype_database.txt&quot;,
         &quot;known_indels_sites_indices&quot; : [ &quot;gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz.tbi&quot;, &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz.tbi&quot; ],
         &quot;evaluation_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/exome_evaluation_regions.v1.interval_list&quot;,
         &quot;contamination_sites_mu&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.mu&quot;,
         &quot;dbsnp_vcf&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf&quot;,
         &quot;known_indels_sites_vcfs&quot; : [ &quot;gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz&quot;, &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz&quot; ],
         &quot;reference_fasta&quot; : {
           &quot;ref_pac&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.pac&quot;,
           &quot;ref_bwt&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.bwt&quot;,
           &quot;ref_dict&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dict&quot;,
           &quot;ref_ann&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.ann&quot;,
           &quot;ref_fasta_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.fai&quot;,
           &quot;ref_alt&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.alt&quot;,
           &quot;ref_fasta&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta&quot;,
           &quot;ref_sa&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.sa&quot;,
           &quot;ref_amb&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.amb&quot;
         },
         &quot;contamination_sites_ud&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.UD&quot;
       },
       &quot;cram_ref_fasta_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.fai&quot;,
       &quot;papi_settings&quot; : {
         &quot;agg_preemptible_tries&quot; : 3,
         &quot;preemptible_tries&quot; : 3
       },
       &quot;base_file_name&quot; : &quot;a.cram&quot;,
       &quot;bait_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/HybSelOligos/whole_exome_illumina_coding_v1/whole_exome_illumina_coding_v1.Homo_sapiens_assembly38.baits.interval_list&quot;,
       &quot;final_gvcf_base_name&quot; : &quot;a.cram&quot;,
       &quot;target_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/HybSelOligos/whole_exome_illumina_coding_v1/whole_exome_illumina_coding_v1.Homo_sapiens_assembly38.targets.interval_list&quot;,
       &quot;unmapped_bam_suffix&quot; : &quot;.unmapped.bam&quot;
     }
   } ],
   &quot;commit&quot; : &quot;commit-ish&quot;,
   &quot;project&quot; : &quot;PO-1234&quot;,  
   &quot;uuid&quot; : &quot;1337254e-f7d8-438d-a2b3-a74b199fee3c&quot;,  
   &quot;wdl&quot; : &quot;pipelines/broad/reprocessing/external/exome/ExternalExomeReprocessing.wdl&quot;,
   &quot;version&quot; : &quot;X.Y.Z&quot;
 }
</code></pre></div></p>
<h3 id="query-workload-apiv1workloaduuiduuid">Query Workload: <code>/api/v1/workload?uuid=&lt;uuid&gt;</code><a class="headerlink" href="#query-workload-apiv1workloaduuiduuid" title="Permanent link">&para;</a></h3>
<p>Queries the WFL database for workloads. Specify the uuid to query for a specific workload.</p>
<p>Request:
<div class="highlight"><pre><span></span><code>curl -X GET &#39;https://dev-wfl.gotc-dev.broadinstitute.org/api/v1/workload?uuid=1337254e-f7d8-438d-a2b3-a74b199fee3c&#39; \
  -H &quot;Authorization: Bearer $(gcloud auth print-access-token)&quot;
</code></pre></div></p>
<p>Response:
<div class="highlight"><pre><span></span><code>[{
   &quot;creator&quot; : &quot;user@domain&quot;,
   &quot;pipeline&quot; : &quot;ExternalExomeReprocessing&quot;,
   &quot;cromwell&quot; : &quot;https://cromwell-gotc-auth.gotc-dev.broadinstitute.org&quot;,
   &quot;release&quot; : &quot;ExternalExomeReprocessing_vX.Y.Z&quot;,
   &quot;created&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
   &quot;started&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
   &quot;output&quot; : &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/&quot;,
   &quot;workflows&quot; : [ {
     &quot;status&quot; : &quot;Submitted&quot;
     &quot;updated&quot; : &quot;YYYY-MM-DDTHH:MM:SSZ&quot;,
     &quot;uuid&quot; : &quot;bb0d93e3-1a6a-4816-82d9-713fa58fb235&quot;,
     &quot;inputs&quot; : {
       &quot;cram_ref_fasta&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta&quot;,
       &quot;scatter_settings&quot; : {
         &quot;haplotype_scatter_count&quot; : 50,
         &quot;break_bands_at_multiples_of&quot; : 0
       },
       &quot;input_cram&quot; : &quot;gs://path/to/a.cram&quot;,
       &quot;destination_cloud_path&quot; : &quot;gs://broad-gotc-dev-wfl-ptc-test-outputs/xx-test-output/to&quot;,
       &quot;sample_name&quot; : &quot;a&quot;,
       &quot;bait_set_name&quot; : &quot;whole_exome_illumina_coding_v1&quot;,
       &quot;references&quot; : {
         &quot;calling_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/exome_calling_regions.v1.interval_list&quot;,
         &quot;contamination_sites_bed&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.bed&quot;,
         &quot;dbsnp_vcf_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf.idx&quot;,
         &quot;haplotype_database_file&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.haplotype_database.txt&quot;,
         &quot;known_indels_sites_indices&quot; : [ &quot;gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz.tbi&quot;, &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz.tbi&quot; ],
         &quot;evaluation_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/exome_evaluation_regions.v1.interval_list&quot;,
         &quot;contamination_sites_mu&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.mu&quot;,
         &quot;dbsnp_vcf&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf&quot;,
         &quot;known_indels_sites_vcfs&quot; : [ &quot;gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz&quot;, &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz&quot; ],
         &quot;reference_fasta&quot; : {
           &quot;ref_pac&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.pac&quot;,
           &quot;ref_bwt&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.bwt&quot;,
           &quot;ref_dict&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dict&quot;,
           &quot;ref_ann&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.ann&quot;,
           &quot;ref_fasta_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.fai&quot;,
           &quot;ref_alt&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.alt&quot;,
           &quot;ref_fasta&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta&quot;,
           &quot;ref_sa&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.sa&quot;,
           &quot;ref_amb&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.amb&quot;
         },
         &quot;contamination_sites_ud&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/contamination-resources/1000g/1000g.phase3.100k.b38.vcf.gz.dat.UD&quot;
       },
       &quot;cram_ref_fasta_index&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.fai&quot;,
       &quot;papi_settings&quot; : {
         &quot;agg_preemptible_tries&quot; : 3,
         &quot;preemptible_tries&quot; : 3
       },
       &quot;base_file_name&quot; : &quot;a.cram&quot;,
       &quot;bait_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/HybSelOligos/whole_exome_illumina_coding_v1/whole_exome_illumina_coding_v1.Homo_sapiens_assembly38.baits.interval_list&quot;,
       &quot;final_gvcf_base_name&quot; : &quot;a.cram&quot;,
       &quot;target_interval_list&quot; : &quot;gs://gcp-public-data--broad-references/hg38/v0/HybSelOligos/whole_exome_illumina_coding_v1/whole_exome_illumina_coding_v1.Homo_sapiens_assembly38.targets.interval_list&quot;,
       &quot;unmapped_bam_suffix&quot; : &quot;.unmapped.bam&quot;
     }
   } ],
   &quot;commit&quot; : &quot;commit-ish&quot;,
   &quot;project&quot; : &quot;PO-1234&quot;,  
   &quot;uuid&quot; : &quot;1337254e-f7d8-438d-a2b3-a74b199fee3c&quot;,  
   &quot;wdl&quot; : &quot;pipelines/broad/reprocessing/external/exome/ExternalExomeReprocessing.wdl&quot;,
   &quot;version&quot; : &quot;X.Y.Z&quot;
 }]
</code></pre></div></p>
<p>The "workflows" field lists out each Cromwell workflow that was started, and
includes their status information. It is also possible to use the Job Manager
to check workflow progress and easily see information about any workflow
failures.</p>
<h2 id="a1-external-exome-workload-request-json-spec">A1 External Exome Workload-Request JSON Spec<a class="headerlink" href="#a1-external-exome-workload-request-json-spec" title="Permanent link">&para;</a></h2>
<div class="highlight"><pre><span></span><code>{
     &quot;pipeline&quot;: &quot;string&quot;,
     &quot;cromwell&quot;: &quot;string&quot;,
     &quot;output&quot;: &quot;string&quot;,
     &quot;project&quot;: &quot;string&quot;,
     &quot;shared_inputs&quot;: {
         &quot;unmapped_bam_suffix&quot;:    &quot;string&quot;,
         &quot;cram_ref_fasta&quot;:       &quot;string&quot;,
         &quot;cram_ref_fasta_index&quot;: &quot;string&quot;,
         &quot;bait_set_name&quot;:        &quot;string&quot;,
         &quot;bait_interval_list&quot;:   &quot;string&quot;,
         &quot;target_interval_list&quot;: &quot;string&quot;
         &quot;references&quot;: {
             &quot;calling_interval_list&quot;:      &quot;string&quot;,
             &quot;contamination_sites_bed&quot;:    &quot;string&quot;,
             &quot;contamination_sites_mu&quot;:     &quot;string&quot;,
             &quot;contamination_sites_ud&quot;:     &quot;string&quot;,
             &quot;dbsnp_vcf&quot;:                  &quot;string&quot;,
             &quot;dbsnp_vcf_index&quot;:            &quot;string&quot;,
             &quot;evaluation_interval_list&quot;:   &quot;string&quot;,
             &quot;haplotype_database_file&quot;:     &quot;string&quot;,
             &quot;known_indels_sites_vcfs&quot;:    [&quot;string&quot;],
             &quot;known_indels_sites_indices&quot;: [&quot;string&quot;],
             &quot;reference_fasta&quot;: {
                 &quot;ref_pac&quot;:         &quot;string&quot;,
                 &quot;ref_bwt&quot;:         &quot;string&quot;,
                 &quot;ref_dict&quot;:        &quot;string&quot;,
                 &quot;ref_ann&quot;:         &quot;string&quot;,
                 &quot;ref_fasta_index&quot;: &quot;string&quot;,
                 &quot;ref_alt&quot;:         &quot;string&quot;,
                 &quot;ref_fasta&quot;:       &quot;string&quot;,
                 &quot;ref_sa&quot;:          &quot;string&quot;,
                 &quot;ref_amb&quot;:         &quot;string&quot;
             }
         },
         &quot;scatter_settings&quot;: {
             &quot;haplotype_scatter_count&quot;:     int,
             &quot;break_bands_at_multiples_of&quot;: int
         },
         &quot;papi_settings&quot;: {
             &quot;agg_preemptible_tries&quot;: int,
             &quot;preemptible_tries&quot;:     int
         },
         &quot;fingerprint_genotypes_file&quot;: &quot;string&quot;,
         &quot;fingerprint_genotypes_index&quot;: &quot;string&quot;,
     },
     &quot;items&quot;: [{
         &quot;inputs&quot;: {
             // required - specify either &quot;input_bam&quot; or &quot;input_cram&quot;
             &quot;input_bam&quot;:  &quot;string&quot;,
             &quot;input_cram&quot;: &quot;string&quot;,

             // optional inputs
             &quot;sample_name&quot;:          &quot;string&quot;,
             &quot;final_gvcf_base_name&quot;:  &quot;string&quot;,
             &quot;unmapped_bam_suffix&quot;:    &quot;string&quot;,
             &quot;cram_ref_fasta&quot;:       &quot;string&quot;,
             &quot;cram_ref_fasta_index&quot;: &quot;string&quot;,
             &quot;bait_set_name&quot;:        &quot;string&quot;,
             &quot;bait_interval_list&quot;:   &quot;string&quot;,
             &quot;target_interval_list&quot;: &quot;string&quot;
             &quot;references&quot;: {
                 &quot;calling_interval_list&quot;:      &quot;string&quot;,
                 &quot;contamination_sites_bed&quot;:    &quot;string&quot;,
                 &quot;contamination_sites_mu&quot;:     &quot;string&quot;,
                 &quot;contamination_sites_ud&quot;:     &quot;string&quot;,
                 &quot;dbsnp_vcf&quot;:                  &quot;string&quot;,
                 &quot;dbsnp_vcf_index&quot;:            &quot;string&quot;,
                 &quot;evaluation_interval_list&quot;:   &quot;string&quot;,
                 &quot;haplotype_database_file&quot;:     &quot;string&quot;,
                 &quot;known_indels_sites_vcfs&quot;:    [&quot;string&quot;],
                 &quot;known_indels_sites_indices&quot;: [&quot;string&quot;],
                 &quot;reference_fasta&quot;: {
                     &quot;ref_pac&quot;:         &quot;string&quot;,
                     &quot;ref_bwt&quot;:         &quot;string&quot;,
                     &quot;ref_dict&quot;:        &quot;string&quot;,
                     &quot;ref_ann&quot;:         &quot;string&quot;,
                     &quot;ref_fasta_index&quot;: &quot;string&quot;,
                     &quot;ref_alt&quot;:         &quot;string&quot;,
                     &quot;ref_fasta&quot;:       &quot;string&quot;,
                     &quot;ref_sa&quot;:          &quot;string&quot;,
                     &quot;ref_amb&quot;:         &quot;string&quot;
                 }
             },
             &quot;scatter_settings&quot;: {
                 &quot;haplotype_scatter_count&quot;:     int,
                 &quot;break_bands_at_multiples_of&quot;: int
             },
             &quot;papi_settings&quot;: {
                 &quot;agg_preemptible_tries&quot;: int,
                 &quot;preemptible_tries&quot;:     int
             },
             &quot;fingerprint_genotypes_file&quot;:  &quot;string&quot;,
             &quot;fingerprint_genotypes_index&quot;: &quot;string&quot;,
             &quot;destination_cloud_path&quot;:      &quot;string&quot;
         }
     }]
}
</code></pre></div>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../dev-wgs/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Whole Genome
              </div>
            </div>
          </a>
        
        
          <a href="../terra/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                WorkFlow Launcher's Role in Terra
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/broadinstitute/wfl" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/vendor.6a3d08fc.min.js"></script>
      <script src="../assets/javascripts/bundle.71201edf.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "..",
          features: [],
          search: Object.assign({
            worker: "../assets/javascripts/worker/search.4ac00218.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>